src/Combinatorics/OrderedMultiIndex.jl

```diff
@@ -183,3 +183,50 @@ function ordered_multi_index(k::Int, p::Int, n::Int)
   end
 end
 
+mutable struct MultiIndicesOfDegree
+  n::Int
+  d::Int
+
+  function MultiIndicesOfDegree(n::Int, d::Int)
+    @assert n >= 0 "length of index must be non-negative"
+    @assert d >= 0 "degree must be non-negative"
+    return new(n, d)
+  end
+end
```

Comment:
@joschmitt : I needed an iterator over all multiindices `I = (i_1, i_2, ..., i_n)` with `i_k >= 0` and `i_1 + i_2 + ... + i_n = d` here. I didn't want to go through your `monomials_of_degree` things, because I would have to create a dummy ring for that and then extract the exponent vectors. 

Question: Are you using such an iterator in the internals of your code? If not, do you think it would make sense to build your iterator on this one?

Reply:
I think in combinatorics this is called $d$-multicombinations of $n$ and yes, it makes sense that there is such an iterator that produces the integer vectors and then one can for example build the monomials out of it.
(I didn't implement it like this back then because I hoped that some kind combinatorics expert would come round and do it properly... For the same reason, I also have my own multiset partitions somewhere in the invariant theory code...😞)

Caveat: Your algorithm appears to be different from what I implemented (for which I don't have a reference, maybe I just took what they have in `Combinatorics.jl`, I don't remember), so I would like to compare these to approaches runtime-wise before we delete "my" implementation. I can also do this once this pull request is merged.

Reply:
Sounds good! I also simply did a straightforward (and probably not very optimal) implementation. But as long as we don't have it, I think it's better to have *some* implementation rather than none. 

Comparing the speed of either of the two is good! Finding a combinatorics expert to tweak our stuff would be even better. But that can also be done later since the user facing part will not change.

Reply:
I don't understand how one gets indices which sum up to a given `d` with this?

Reply:
Probably it is fastest if you just compute the `partitions` of your number of the right size(s) and then extend via group action.

Reply:
Nicely done! 

In principle you should be able to push directly to this PR if you check it out with `gh pr checkout 3021`. If you run the tests for the modules and those in `experimental/DoubleAndHyperComplexes/test/` before you push, then I think it's fine if you pushed your work here.

But it's also fine with me to do that later.